---
title: "Downstream analysis of Cholangiocarcinoma genomics and clinical data"
author: "Nasim Bondar Sahebi"
date: "2023-12-20"
output:
  html_document: default
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```


# Load libraries.

```{r  load libraries}

library(SummarizedExperiment)
library(dplyr)
library(ggplot2)
library(knitr)
library(DT)
library(limma)
library(EnhancedVolcano)

```


# Data Loading and Preparation

## Load Gene Summarized Experiment Data

#### Data Overview

  - **Paper name **:  Genomic characterization of cholangiocarcinoma with autoimmune etiologies
  - **Patient samples **: 33 
  - **RNA sequencing (RNA-seq)**: 33 patients
  - **Non-autoimmune **:  25 patients
  - **Non-autoimmune **:  15 patients
  

Load the Summarized Experiment .rds file (SE_CCA.rds), extract clinical and expression data (TPM), along with annotations, and prepare the gene expression data for analysis.

```{r summery load-data}

# Load your SE result and extract clinical data , expression data and annotation

# Load SE obj
se <- readRDS("~/BHK lab/kevin Project/Cholangiocarcinoma/data/SE_CCA.rds")

# Extract Clinical data 
clin <- data.frame(colData(se)) #dim is 33 x 11

# Extract the expression data
expr_tpm <- assays(se)[["expr_tpm"]] #dim 38040 x 33

# Extract the annotation
annot <- data.frame(rowData(se))  #dim is 38040 x 11

# Display first few rows of the dataset
DT::datatable(round(expr_tpm[1:8, 1:8], 3)) # round to three digits for visualization

```

# Data Preparation 

Preparing expression data (RNA-seq) for analysis:

1. **Restrict expression data to protein-coding genes.**
2. **Remove low/zero expression genes.**


```{r Data Preparation }

# Step 1: Restrict to Protein-Coding Genes.
annot_proteincoding <- annot[annot$gene_type == "protein_coding",] # 19205 protein coding genes.
expr_tpm<- expr_tpm[rownames(expr_tpm) %in% rownames(annot_proteincoding),]
dim(expr_tpm)

# Step 2: Filter Low/Zero Expression Genes]

#data is log2(TPM+1)
r <- as.numeric(apply(expr_tpm, 1, function(i) sum(round(i, 6) == round(log2(1), 6))))

# Get the indices of rows to remove
remove <- which(r > dim(expr_tpm)[2] * 0.5) 

# Remove rows from the matrix
expr_tpm <- expr_tpm[-remove, ]   #dim is 16256 x 33

dim(expr_tpm)

DT::datatable(round(expr_tpm[1:8, 1:8], 3))
```

# Descriptive analysis

## Evaluate Differences in Clinicodemographic and Staging Variables

- **Continuous Variables (e.g., Age):** We calculated the median and IQR and assessed associations using the Wilcoxon test. 

- **Discrete/Categorical Variables (e.g., Sex, Stage, Location):** We determined frequency and percentage distributions and assessed associations using Fisher's exact test.


```{r Descriptive analysis}


# Set 'y' and 'n' to 'Autoimmune' and 'Non-Autoimmune' within the dataframe
clin$psc_ibd[clin$psc_ibd == 'y'] <- 'Autoimmune'
clin$psc_ibd[clin$psc_ibd == 'n'] <- 'Non-Autoimmune'

# 1. Age

# Summarize age statistics (count,Mean(SD), Median(IQR), Min,Max) for each psc_ibd group.
summary_stats <- summarise(
  group_by(clin, psc_ibd),
  count = n(),
  Mean_SD = sprintf("%.2f(%.2f)", mean(age, na.rm = TRUE), sd(age, na.rm = TRUE)),
  Median_IQR = sprintf("%.2f(%.2f)", median(age, na.rm = TRUE), IQR(age, na.rm = TRUE)),
  Min_Max = sprintf("%.2f,%.2f", min(age, na.rm = TRUE), max(age, na.rm = TRUE))
)

# Store statistics for Auto and Non-Auto for later use
auto_count <- summary_stats$count[summary_stats$psc_ibd == "Autoimmune"]
non_auto_count <- summary_stats$count[summary_stats$psc_ibd == "Non-Autoimmune"]

# Visualize your data.

# Plotting boxplot for age distributions
ggplot(clin, aes(x = psc_ibd, y = age, color = psc_ibd)) +
  geom_boxplot() +
  scale_color_manual(values = c("Autoimmune" = "#00AFBB", "Non-Autoimmune" = "#E7B800")) +
  labs(y = "Age", x = "PSC-IBD")

# Perform two-sample Wilcoxon test
res <- wilcox.test(age ~ psc_ibd, data = clin, exact = FALSE)
p_value_age <- round(res$p.value, 2)

# 2. Sex

# Perform Fisher's Exact Test for 'sex' and 'psc_ibd'
contingency_sex <- table(clin$sex, clin$psc_ibd)
p_value_sex <- round(fisher.test(contingency_sex)$p.value,2)

# Calculate and format percentages for Sex
sex_stats <- apply(contingency_sex, 2, function(x) sprintf("%d (%.1f%%)", x, 100 * x / sum(x)))

# Perform Fisher's Exact Test for 'location' and 'psc_ibd'
contingency_loc <- table(clin$location, clin$psc_ibd)
p_value_loc <- round(fisher.test(contingency_loc)$p.value, 2)

# Calculate and format percentages for Location
location_stats <- apply(contingency_loc, 2, function(x) sprintf("%d (%.1f%%)", x, 100 * x / sum(x)))

# Perform Fisher's Exact Test for 'stage' and 'psc_ibd'
# Convert "IIIA", "IIIB", "IIIC" to "III and convert "IVA", "IVB" to "IV".
clin$stage <- with(clin, ifelse(stage %in% c("IIIA", "IIIB", "IIIC"), "III",
                                ifelse(stage %in% c("IVA", "IVB"), "IV", stage)))

contingency_stage <- table(clin$stage, clin$psc_ibd)
p_value_stage <- round(fisher.test(contingency_stage)$p.value, 2)


# Calculate and format percentages for Stage
stage_stats <- apply(contingency_stage, 2, function(x) sprintf("%d (%.1f%%)", x, 100 * x / sum(x)))


# Combine all statistics into a data frame for the table
result <- data.frame(
  Characteristics = c("Age", "Mean (SD)", "Median (IQR)", "Min, Max", "",
                      "Sex", "F", "M", "",
                      "Location", "dCCA", "GBC", "iCCA", "pCCA", "",
                      "Stage", "II", "III", "IV"),
  
  `Autoimmune (n=14)` = c("", summary_stats$Mean_SD[1], summary_stats$Median_IQR[1], summary_stats$Min_Max[1], "","",
                          sex_stats[1, 1], sex_stats[2, 1], "", "", location_stats[1, 1], location_stats[2, 1],
                          location_stats[3, 1], location_stats[4, 1],"","", stage_stats[1, 1], stage_stats[2, 1], 
                          stage_stats[3, 1]),
  
  `Non-Autoimmune (n=19)` = c("", summary_stats$Mean_SD[2], summary_stats$Median_IQR[2], summary_stats$Min_Max[2], "","",
                              sex_stats[1, 2], sex_stats[2, 2], "", "", location_stats[1, 2], location_stats[2, 2],
                              location_stats[3, 2], location_stats[4, 2],"","", stage_stats[1, 2], stage_stats[2, 2],
                              stage_stats[3, 2]),
  
  `P value` = c(p_value_age, "", "", "", "",
                p_value_sex, "", "", "", p_value_loc,
                "", "", "", "", "",
                p_value_stage, "", "", ""),
  
  check.names = FALSE
)



table_html <- kable(
    result, 
    format = "html", 
    caption =" Table 1: Clinicodemographic and Staging Characteristics at Baseline (AJCC 8th Edition)", 
    escape = FALSE,
    table.attr = 'style="width:100%; overflow-x:scroll; overflow-y:scroll; max-height:400px;"'
)

table_html

```


# Differential gene expression.

  - **Use limma approach in R **: To find the differentially expressed genes between Autoimmune and Non-Autoimmune.
  - **P-values are corrected to control false discovery rate (FDR) using Benjamini-Hochberg approach **
  - ** out put: Limma_Fit.txt file **
  

```{r diff expression }

# Limma approach applied.
design <- model.matrix(~ clin$psc_ibd)
fit <- lmFit(expr_tpm, design)
fit <- eBayes(fit)

# Sort by Pvalue.
top.table <- topTable(fit, sort.by = "P", n = Inf)

# Preview
DT::datatable(top.table)

# To see How many DE gene are there.
length(which(top.table$adj.P.Val < 0.05))  # 460 DE genes.

# Save top table to .txt file.
top.table$Gene <- rownames(top.table)
write.table(top.table, file = "~/BHK lab/kevin Project/Cholangiocarcinoma/data/Limma_Fit.txt", row.names = F, sep = "\t", quote = F)


```

## Volcano plot

 - ** Volcano Plot based on P value and F **


```{r  Volcano plot libraries,fig.width=10, fig.height=10}


# 1. Volcano plot.

# 1.1 Volcano Plot based on P value
EnhancedVolcano(top.table,
    lab = rownames(top.table),
    x = 'logFC',
    y = 'P.Value',
    title= 'Volcano Plot based on P value'
    )

# 1.2 Volcano Plot based on FDR 
# adjusted p-values as red 
EnhancedVolcano(top.table,
    lab = rownames(top.table),
    x = 'logFC',
    y = 'adj.P.Val',
    pCutoff = 0.05,
    FCcutoff = 1.5,
    xlim = c(-5, 5),
    legendLabels=c('Not sig.','Log FC','adj p-value',
      'adj p-value & Log FC'),
    ylim = c(0, -log10(10e-4)),
    title= 'Volcano Plot based on FDR',)


```